Setup fsGroup for local volumes correctly

cofyc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This pr fixes fsGroup check in local volume in containerized kubelet. Except this, it also fixes fsGroup check when volume source is a normal directory whether kubelet is running on the host or in a container.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61741 

**Special notes for your reviewer**:

Bind mounts are detected in `/proc/mounts`, but it does not contain root of mount for bind mounts. So `mount.GetMountRefsByDev()` cannot get all references if source is a normal directory. e.g.

```
# mkdir /tmp/src /mnt/dst
# mount --bind /tmp/src /tmp/src # required by local-volume-provisioner, see https://github.com/kubernetes-incubator/external-storage/pull/499
# mount --bind /tmp/src /mnt/dst
# grep -P 'src|dst' /proc/mounts 
tmpfs /tmp/src tmpfs rw,nosuid,nodev,noatime,size=4194304k 0 0
tmpfs /mnt/dst tmpfs rw,nosuid,nodev,noatime,size=4194304k 0 0
# grep -P 'src|dst' /proc/self/mountinfo 
234 409 0:42 /src /tmp/src rw,nosuid,nodev,noatime shared:30 - tmpfs tmpfs rw,size=4194304k
235 24 0:42 /src /mnt/dst rw,nosuid,nodev,noatime shared:30 - tmpfs tmpfs rw,size=4194304k
```

We need to compare root of mount and device in this case.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

cofyc commented:
/assign @msau42

msau42 commented:
This is a pretty big change to just fix local volumes.  Can you add this as a new interface instead and keep existing Mounter interfaces using /proc/mounts?

/assign @jsafrane

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -265,6 +265,34 @@ var _ = utils.SIGDescribe("PersistentVolumes-local ", func() {
+					Expect(err).NotTo(HaveOccurred(), "failed: getting the right privileges in the file %v", int(fsGroup))
+				})
+
+				It("should not set fsGroup twice for local volume", func() {
```
From the original fsgroup [PR](https://github.com/kubernetes/kubernetes/pull/49988), it was determined that we can't actually enforce that the 2nd pod mount will fail due to different fsgroup.  The best we could do is log a warning.

cofyc replied:
Logs are updated to be more correct. https://github.com/kubernetes/kubernetes/pull/62102/commits/7b1ec6f0eef7a96f97eb3c896f63c3a56bb02d5e

If local volume is already mounted by a pod, [`fsGroup` will be not enforced](https://github.com/kubernetes/kubernetes/pull/49988/files#diff-ec924773446516c2e7bf84c477f5f4ddR288). The second e2e test is used to check this. 

The local plugin [emits a warning event](https://github.com/kubernetes/kubernetes/pull/49988/files#diff-ec924773446516c2e7bf84c477f5f4ddR241), when `fsGroup` cannot be enforced. But event is [only logged by `glog`](https://github.com/kubernetes/kubernetes/pull/49988/files#diff-ec924773446516c2e7bf84c477f5f4ddR60), not reported to apiserver, I cannot verified it in e2e tests. Should we report this event to apiserver, then user can know it by `describe pod`?

msau42 replied:
The intent was for the event to be written to API server.  So that's another bug :(

cofyc replied:
@msau42 OK, I will fix it in another PR.

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -1059,6 +1087,30 @@ func createLocalPod(config *localTestConfig, volume *localTestVolume) (*v1.Pod,
 	return framework.CreateSecPod(config.client, config.ns, []*v1.PersistentVolumeClaim{volume.pvc}, false, "", false, false, selinuxLabel)
 }
 
+func createLocalPodWithFsGroup(config *localTestConfig, volume *localTestVolume, fsGroup *int64) (*v1.Pod, error) {
```
`MakeSecPod` already sets an fsgroup, so we shouldn't need this

cofyc replied:
@msau42 
`MakeSecPod` use a fixed `fsGroup` to create pod, but I need to different `fsGroup` for each pod. I cannot use it. So I need this function.

Another way is to extend `MakeSecPod` and `CreateSecPod` to accept `fsGroup` parameter. Then I can reuse `MakeSecPod`. What do you think?

msau42 replied:
Yeah, reusing is fine.

cofyc replied:
@msau42 Done in https://github.com/kubernetes/kubernetes/pull/62102/commits/2b3bdd1f6ba36b84962f30f37dc21e7d70c26f98.

msau42 commented on pkg/util/mount/mount.go:
```diff
@@ -166,19 +183,21 @@ func (mounter *SafeFormatAndMount) FormatAndMount(source string, target string,
+// mountSource could be a mountpoint path, device or a normal directory (for
+// bind mount).
+func GetMountRefsBySource(mounter Interface, mountSource string) ([]string, error) {
 	mps, err := mounter.List()
```
Would prefer having a new List method that uses mountinfo instead of changing existing List

cofyc replied:
@msau42 Done in https://github.com/kubernetes/kubernetes/pull/62102/commits/e710fa51e1631090f9c0913765b58a94ee1f8e6a.

I don't find any downside to extend `MountPoint` struct to include information from `/proc/<pid>/mountinfo`, except it exposes more linux-specific details. However, I agree this is unnecessary to fix the issue. IMHO, maybe we can unify them after more tests on various platform.

cofyc commented on pkg/util/mount/mount.go:
```diff
@@ -84,6 +84,8 @@ type Interface interface {
 	// Will operate in the host mount namespace if kubelet is running in a container
 	MakeDir(pathname string) error
+	// TruePath returns true path which kubelet can access whether it is running in a container or not
+	TruePath(pathname string) string
```
@msau42 Done.

msau42 commented on pkg/volume/local/local.go:
```diff
@@ -285,7 +285,7 @@ func (m *localVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
 		if len(refs) > 0 {
 			fsGroupNew := int64(*fsGroup)
-			fsGroupSame, fsGroupOld, err := volume.IsSameFSGroup(m.globalPath, fsGroupNew)
+			fsGroupSame, fsGroupOld, err := volume.IsSameFSGroup(m.mounter.KubeletPath(m.globalPath), fsGroupNew)
```
@jsafrane pointed out a problem with using a wrapped path in a [different PR](https://github.com/kubernetes/kubernetes/pull/61489#pullrequestreview-110032850).  So I think we need to pass in `mounter` and invoke `stat` on the host instead

msau42 commented on pkg/util/mount/mount.go:
```diff
@@ -166,34 +198,56 @@ func (mounter *SafeFormatAndMount) FormatAndMount(source string, target string,
 		return nil, err
 	}
-	slTarget, err := filepath.EvalSymlinks(mountPath)
+	slTarget, err := filepath.EvalSymlinks(source)
```
I think this EvalSymlinks won't work for the same reason too.

cofyc replied:
@msau42 I proposed a solution here: https://github.com/kubernetes/kubernetes/pull/61489#discussion_r180301910

msau42 commented on pkg/util/mount/mount.go:
```diff
@@ -84,6 +87,11 @@ type Interface interface {
+	KubeletPath(pathname string) string
+	// GetRootPathAndMajorMinor returns root path relative to its mountpoint and
+	// major:minor of filesystem containing given path.
+	GetRootPathAndMajorMinor(pathname string) (string, string, error)
```
Do we need this interface? Isn't the information in the MountInfo struct?

cofyc replied:
@msau42 

It's possible to have same root paths in list of MountInfos. I wrote this interface to find right mount entry by comparing root path relative to its mountpoint and device major:minor in list of MountInfos. There are two cases, if they are self bind mounted, we don't need this interface. 

### If directories is not self bind mounted, this interface is needed

```
# two different filesystems 
mkdir -p /mnt/tmpfs{1,2}
mount -t tmpfs tmpfs /mnt/tmpfs1
mount -t tmpfs tmpfs /mnt/tmpfs2
# two directorys with same path in different filesystems
mkdir -p /mnt/tmpfs{1,2}/vol
# mount to another place to use
mkdir -p /mnt/dst{1,2}
mount --bind /mnt/tmpfs1/vol /mnt/dst1
mount --bind /mnt/tmpfs2/vol /mnt/dst2
#  grep -P '/mnt/dst(1|2)' /proc/self/mountinfo
62 24 0:84 /vol /mnt/dst1 rw,relatime shared:45 - tmpfs tmpfs rw
67 24 0:85 /vol /mnt/dst2 rw,relatime shared:46 - tmpfs tmpfs rw
#  grep -P '/mnt/tmpfs(1|2)' /proc/self/mountinfo
59 24 0:84 / /mnt/tmpfs1 rw,relatime shared:45 - tmpfs tmpfs rw
61 24 0:85 / /mnt/tmpfs2 rw,relatime shared:46 - tmpfs tmpfs rw
```

### If directories is self bind mounted, this interface is unnecessary
 
```
# two different filesystems 
mkdir -p /mnt/tmpfs{1,2}
mount -t tmpfs tmpfs /mnt/tmpfs1
mount -t tmpfs tmpfs /mnt/tmpfs2
# two directorys with same path in different filesystems
mkdir -p /mnt/tmpfs{1,2}/vol
# self-bind
mount --bind /mnt/tmpfs1/vol /mnt/tmpfs1/vol
mount --bind /mnt/tmpfs2/vol /mnt/tmpfs2/vol
# mount to another place to use
mkdir -p /mnt/dst{1,2}
mount --bind /mnt/tmpfs1/vol /mnt/dst1
mount --bind /mnt/tmpfs2/vol /mnt/dst2
# grep -P '/mnt/dst(1|2)' /proc/self/mountinfo
68 24 0:84 /vol /mnt/dst1 rw,relatime shared:45 - tmpfs tmpfs rw
69 24 0:85 /vol /mnt/dst2 rw,relatime shared:46 - tmpfs tmpfs rw
# grep -P '/mnt/tmpfs(1|2)' /proc/self/mountinfo
59 24 0:84 / /mnt/tmpfs1 rw,relatime shared:45 - tmpfs tmpfs rw
61 24 0:85 / /mnt/tmpfs2 rw,relatime shared:46 - tmpfs tmpfs rw
62 59 0:84 /vol /mnt/tmpfs1/vol rw,relatime shared:45 - tmpfs tmpfs rw
67 61 0:85 /vol /mnt/tmpfs2/vol rw,relatime shared:46 - tmpfs tmpfs rw
```

We can find root path ("/vol") and major:minor ("0:84") by path`/mnt/tmpfs1/vol` in mount entry 62. Should we assume directory volumes are always self bind mounted?

msau42 replied:
Thanks for the explanation.  You are right, we cannot assume directory volumes are always self bind mounted.

I still have a question about why we need an interface method for this though.  `ListMountInfos` will return every entry in /proc/self/mountinfo, and it will contain the information about the mount root and device numbers.  Can you explain why in `MountGetRefsBySource`, we need to call this `GetRootPath` method when the `ListMountInfos` should already have all the information we need?  Instead of an interface, can `GetRootPath` just be a helper method?

cofyc replied:
> Can you explain why in MountGetRefsBySource, we need to call this GetRootPath method when the ListMountInfos should already have all the information we need?

Because  in `/proc/self/mountinfo`, bind mount entry for `/mnt/dst1` does not use absolute path (`/mnt/tmpfs1/vol`) as mount source, it only contains its root path (`/vol`) and filesystem information (`0:84`), e.g.

```
62 24 0:84 /vol /mnt/dst1 rw,relatime shared:45 - tmpfs tmpfs rw
```

And same path can exist on more than one filesystems. `/mnt/tmpfs1/vol`, `/mnt/tmpfs2/vol` have same root path `/vol` under their filesystems. So for directory bind mounting, we need to use root path (path relative to its fs root) and filesystem (device `major:minor`) to uniquely identify bind source.

In case 1,  we cannot find its root path and filesystem in `/proc/self/mountinfo`, because no mount entries contains information for `/mnt/tmpfs1/vol`, because its actually a normal directory. So we need `GetRootPathAndMajorMinor()` to find out information we need by accessing its filesystem.

If case 2, if it's self bind mounted, there exists a mount entry 62, from which we known filesystem (`0:84`) and its path (`/vol`) is mounted at `/mnt/tmpfs1/vol`. So we can know `/mnt/tmpfs1/vol` is also mounted at `/mnt/dst1` by mount entry 68, because they have same filesystem (`0:84`) and mount root path (`/vol`).

> Instead of an interface, can GetRootPath just be a helper method?

OK. I'll change that.

cofyc commented:
/test pull-kubernetes-e2e-gce

cofyc commented:
Due to conflicts, I rebased on latest master.

sjenning commented:
/uncc

msau42 commented on pkg/util/mount/mount.go:
```diff
@@ -150,6 +158,30 @@ type MountPoint struct {
+	// Per-mount options.
+	Opts []string
+	// Per-superblock options.
+	SuperOpts []string
```
Do we still need this?  Is this going to break anybody that was expecting the option to be in Opts?

cofyc replied:
`MountInfo` is a separate struct now, I think anybody trying to use it needs to know what fields are first, it will not break anybody IMO.

`Opts` and `SuperOpts` are not used currently. I parsed all fields out from `mountinfo` according to [its doc](http://man7.org/linux/man-pages/man5/proc.5.html) in case future needs arise.

If you prefer, I can remove this fields. What do you think?

cofyc commented:
@msau42 

hi, in https://github.com/kubernetes/kubernetes/pull/62102/commits/3337cb18f5c3ebcb1d493afc57b2de99cd543e85 I changed `KubeletPath` to return path name after the evaluation of any symbolic links in addition to the path name kubelet can access. This fixed issues [here](https://github.com/kubernetes/kubernetes/pull/62102#discussion_r180223662) and [here](https://github.com/kubernetes/kubernetes/pull/62102#discussion_r180223141). I also updated e2e tests to test when volume path is a symbolic link to a normal directory.

After doing a lot of tests, I found main problem of https://github.com/kubernetes/kubernetes/pull/62102 and https://github.com/kubernetes/kubernetes/pull/61489 is to get path name ofter evaluation of symbolic links in container which we need to search mount entry in `/proc/self/mountinfo` and get path kubelet can access in container (for `volume.IsSameFsGroup`, etc).

In this PR, I wrote `KubeletPath` interface to address this problem. It's implemented with [my proposed solution](https://github.com/kubernetes/kubernetes/pull/61489/files#r180301910) to test if it's going to work.

cofyc commented:
/test pull-kubernetes-verify

jsafrane reviewed (commented):
This PR is indeed too complicated. I would suggest to split it into two - one with `mounter.KubeletPath()` to fix just the check in local volume. And second with `GetMountRefsBySource()` that makes non-device as local volumes working.

cofyc commented:
@jsafrane Thanks! I will follow your suggestion to split this into two prs. Add `GetMountRefsBySource` to `mount.Interface` seems a good idea too!

Commit:
Fixes fsGroup check in local volume in containerized kubelet. Except
this, it also fixes fsGroup check when volume source is a normal
directory whether kubelet is running on the host or in a container.

cofyc commented:
/test pull-kubernetes-e2e-gce 
/test pull-kubernetes-e2e-kops-aws

cofyc commented:
@msau42 @jsafrane 
hi, this PR has been updated to fix skipped tests in https://github.com/kubernetes/kubernetes/pull/62903. Please take a look at again.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1146,3 +1198,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
+	last_stat := info.Sys().(*syscall.Stat_t)
+	// Walk up parents until we find another file system.
+	for {
+		upDir := filepath.Dir(path)
```
@jsafrane It will break because `filepath.Dir("/")` is `/`, same `stat.Ino`.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -791,6 +813,36 @@ func (mounter *Mounter) CleanSubPaths(podDir string, volumeName string) error {
+	return fmt.Sprintf("%d:%d", unix.Major(dev), unix.Minor(dev))
+}
+
+func getRootPathAndMarjorMinor(pathname string) (string, string, error) {
```
Done.

cofyc commented on pkg/util/mount/mount.go:
```diff
@@ -49,6 +49,9 @@ type Interface interface {
 	List() ([]MountPoint, error)
+	// ListMountInfos returns a list of mount infos. Like List(), but each
+	// entry contains more information.
+	ListMountInfos() ([]MountInfo, error)
```
Done in https://github.com/kubernetes/kubernetes/pull/62903.

cofyc commented on pkg/util/mount/evalsymlinks_linux.go:
```diff
@@ -0,0 +1,127 @@
+// mountpoint. It works like filepath.EvalSymlinks except it does not assume
+// filesystem root is OS root "/".
+// `filepath.EvalSymlinks` equals to `EvalSymlinksAtNewRoot(path, "/", "/")`.
+func EvalSymlinksAtNewRoot(path string, oldroot, newroot string) (string, error) {
```
Done in https://github.com/kubernetes/kubernetes/pull/62903.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1146,3 +1198,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
+		return
+	}
+	if !info.Mode().IsDir() {
+		// If it's not a directory, use its directory.
```
Fixed now, see https://github.com/kubernetes/kubernetes/pull/62102/files#diff-f0ba2b2ac6f7b574258c97a4001460b2R1294.

cofyc commented:
/test pull-kubernetes-local-e2e-containerized
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-kubemark-e2e-gce

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -611,19 +613,40 @@ func isShared(path string, filename string) (bool, error) {
-	// list of "optional parameters", mount propagation is one of them
-	optional []string
+	// Unique ID for the mount (maybe reused after umount).
+	ID int
```
The struct is non-public, so I simply let fields to be public.

cofyc replied:
Done.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -642,17 +665,38 @@ func parseMountInfo(filename string) ([]mountInfo, error) {
 		fields := strings.Fields(line)
-		if len(fields) < 7 {
-			return nil, fmt.Errorf("wrong number of fields in (expected %d, got %d): %s", 8, len(fields), line)
+		if len(fields) < expectedAtLeastNumFieldsPerMountInfo {
```
Yes, I'm sure. See docs from http://man7.org/linux/man-pages/man5/proc.5.html) and util-linux libmount parser (https://github.com/karelzak/util-linux/blob/v2.32/libmount/src/tab_parse.c#L176).

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1215,109 @@ func doSafeOpen(pathname string, base string) (int, error) {
+	majorMinor := ""
+
+	// Finding the underlying root path and major:minor if source is a mountpoint.
+	for i := range mis {
```
@msau42 Oh, never thought about it. Fixed in https://github.com/kubernetes/kubernetes/pull/62102/commits/ae74b995747cb325f99f61703351775f79597d45.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1215,111 @@ func doSafeOpen(pathname string, base string) (int, error) {
+	}
+
+	// Get root path and major:minor from filesystem if we could not get from mounts.
+	if rootPath == "" || majorMinor == "" {
```
When we use non-mountpoint mount source, e.g. `/dev/sda1` or `/mnt/disks/vol1/subdir`, we cannot detect the `major:minor` and root path from mounts, so we need to access filesystem to retrieve `major:minor` and root path.

`getRootPathAndMajorMinor` is used to get `major:minor` and root path. In these cases:

- `getRootPathAndMajorMinor("/dev/sda1")` will return `<major>:<minor>` of `/dev/sda1` device and root path `/`.
- `getRootPathAndMajorMinor("/mnt/disks/vol1/subdir")` will return `<major>:<minor>` of filesystem mounted at `/mnt/disks/vol1/` and root path `/subdir`.

cofyc commented on pkg/util/mount/mount_linux_test.go:
```diff
@@ -1789,3 +1890,122 @@ func isOperationNotPermittedError(err error) bool {
+	}
+}
+
+func TestFindMountPointByPath(t *testing.T) {
```
Done.

msau42 replied:
Can we test this using a fake /proc/mountinfo file instead?  I don't think we can actually do any real mounting in unit tests.

cofyc commented:
/test pull-kubernetes-e2e-gce

cofyc commented:
/test pull-kubernetes-e2e-gce

jsafrane reviewed (commented):
Mount is very powerful tool and it's hard to parse /proc/self/mountinfo correctly with all its nuances.

`searchMountPoints` is used only by `GetMountRefs`, `GetMountRefs` is used only by `<volume_plugin>.GetDeviceMountRefs`, called when a global mount is being unmounted. Can we use this information to simplify the implementation? For example, we can probably assume the `source` can't be a device. It should be a directory, most probably a mount point, or a file.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1215,111 @@ func doSafeOpen(pathname string, base string) (int, error) {
+// it's possible to mount a non-root path of a filesystem, so we need to use
+// root path and major:minor to represent mount source uniquely.
+// This implementation is shared between Linux and NsEnterMounter
+func searchMountPoints(source, kubeletpath string) ([]string, error) {
```
Fixed in https://github.com/kubernetes/kubernetes/pull/62102/commits/0ccabba1cab295840de8307baa1f95d0667197d5.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -611,19 +613,40 @@ func isShared(path string, filename string) (bool, error) {
+}
+
+func listMountInfos() ([]mountInfo, error) {
+	return parseMountInfo(procMountInfoPath)
```
Fixed in https://github.com/kubernetes/kubernetes/pull/62102/commits/0ccabba1cab295840de8307baa1f95d0667197d5.

cofyc commented:
> searchMountPoints is used only by GetMountRefs, GetMountRefs is used only by > <volume_plugin>.GetDeviceMountRefs, called when a global mount is being unmounted. Can we use this information to simplify the implementation? For example, we can probably assume the source can't be a device. It should be a directory, most probably a mount point, or a file.

hi, @jsafrane 
Actually, there are two `GetMountRefs` functions:

- `mount.GetMountRefs`: this is the function used only by `<volume_plugin>.GetDeviceMountRefs`
- `<mounter>.GetMountRefs` (previously `mount.GetMountRefsByDev` introduced in [#49988](https://github.com/kubernetes/kubernetes/pull/49988/files#diff-0c0020e71c995790a90ad9c61ede7632R154), moved to `Mounter` interface in #62903)

This second function is [used by local volume plugin](https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.0/pkg/volume/local/local.go#L277) to get mount references of local volume source, which can be  a mountpoint, normal directory, symlink, device, etc. So we cannot assume source can't be device. Previous implementation cannot handle some cases, e.g. normal directory. This pr is used to fix this.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1215,111 @@ func doSafeOpen(pathname string, base string) (int, error) {
+
+	// Get root path and major:minor from filesystem if we could not get from mounts.
+	if rootPath == "" || majorMinor == "" {
+		rootPath, majorMinor, err = getRootPathAndMajorMinor(kubeletpath)
```
Good idea! Probably `getRootPathAndMajorMinor` is not necessary actually, because we can get mount references of device by comparing it with `source` field of mount, like it did previously.

cofyc replied:
Updated in https://github.com/kubernetes/kubernetes/pull/62102/commits/ce9261eb64d2c3086b3d0bcc41d02cf2da7d0e3e. `getRootPathAndMajorMinor` and related functions are removed. Code is much simpler.

jsafrane commented:
`mount.GetMountRefs` vs `<mounter>.GetMountRefs`: this is really confusing and one of these should be renamed/removed (in a separate PR).

> Edit: Local volume source can be device, but if source is device volumeMode will be Block <volume_plugin> will not call <mount>.GetMountRefs on it.

Even with https://github.com/kubernetes/kubernetes/pull/63011 merged? It allows FS on block.

cofyc commented:
@jsafrane In #63011, if PV `volumeMode` is `Filesystem` but path is block, [it format and mount block at a global path before using it](https://github.com/kubernetes/kubernetes/pull/63011/files#diff-ec924773446516c2e7bf84c477f5f4dd). So I think we can assume source cannot be device with #63011 merged.

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

jsafrane commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1211,57 @@ func doSafeOpen(pathname string, base string) (int, error) {
+			rootPath = mis[i].root
+			majorMinor = mis[i].majorMinor
+			break
+		} else if strings.HasPrefix(hostSource, mis[i].mountPoint) {
```
Now that I see this, is `HasPrefix` safe? What if:

`source = "/mnt/foobar/local/dir"`
`mis[i].mountPoint = "/mnt/foo"`

mountPoint is prefix of source, but source does not reside on it. `pathWithinBase` might be better option. Please check. It will slow everything down -  how often is  searchMountPoints called?

(and `isShared()` is wrong too, I'll fix it)

cofyc replied:
Not safe. Updated to use `pathWithinBase` in https://github.com/kubernetes/kubernetes/pull/62102/commits/c8007132b3d9cea50fecc63ead8c56d1c925b39f. `searchMountPoints` is called once in `<volume_plugin>.SetUpAt` , performance should be good.

cofyc commented:
/test pull-kubernetes-e2e-gce

jsafrane reviewed (commented):
Now it looks much better, thanks for refactoring.

It should work well when `source` is not a device. Should be `searchMountPoints` renamed to capture this?

If we need to add source as device (which is IMO necessary in future), I would add it as a separate function, say `searchDeviceMounts`

Can you blindly go through all mountInfos and return all of them that have `source` as `mountinfo.source`? Or you can `stat(source)` and use `Stat_t.Rdev` as major/minor.

GetMountRefs would then return `searchMountPoints`+`searchDeviceMounts`.

cofyc commented:
> If we need to add source as device (which is IMO necessary in future), I would add it as a separate function, say searchDeviceMounts
> 
> Can you blindly go through all mountInfos and return all of them that have source as mountinfo.source? Or you can stat(source) and use Stat_t.Rdev as major/minor.
> 
> GetMountRefs would then return searchMountPoints+searchDeviceMounts.

It should work, how about supporting device path as source when necessary in future?

jsafrane commented:
> It should work, how about supporting device path as source when necessary in future?

Sure.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1211,57 @@ func doSafeOpen(pathname string, base string) (int, error) {
+	// We need search in backward order because it's possible for later mounts
+	// to overlap earlier mounts.
+	for i := len(mis) - 1; i >= 0; i-- {
+		if hostSource == mis[i].mountPoint {
```
After some thoughts, you're right, found it's possible to merge these two cases, updated in https://github.com/kubernetes/kubernetes/pull/62102/commits/ef50df7f933a9c349ce4eb4678be31e71b15c7eb.

Fundamentally, we need to find out the filesystem (`major:minor`) and root path of source under that filesystem. For both cases, we can simply join `root` of the filesystem and relative path of source from mount point.

jsafrane replied:
Hmmm, I think I can break the logic with two bind-mounts:

I have tmpfs ("device") on /mnt/data:
```
mkdir /mnt/data
mount -t tmpfs data /mnt/data

# There are two directories in it: /mnt/data/vol1 and /mnt/data/vol2.
mkdir -p /mnt/data/vol{1,2}

# Now bind-mount these directories with magic subdirectories in them:
mkdir /mnt/vol{1,2}
mount --bind /mnt/data/vol1 /mnt/vol1
mount --bind /mnt/data/vol2 /mnt/vol2

mkdir -p /mnt/vol1/foo
mkdir -p /mnt/vol2/vol1/foo

# mount /mnt/vol1/foo to a pod:

mkdir /mnt/pod1
mount --bind /mnt/vol1/foo /mnt/pod1
```


mountinfo:
```
150 60 0:38 / /mnt/data rw,relatime shared:100 - tmpfs data rw,seclabel
154 60 0:38 /vol1 /mnt/vol1 rw,relatime shared:100 - tmpfs data rw,seclabel
158 60 0:38 /vol2 /mnt/vol2 rw,relatime shared:100 - tmpfs data rw,seclabel
162 60 0:38 /vol1/foo /mnt/pod1 rw,relatime shared:100 - tmpfs data rw,seclabel
```

And I want to check references to `/mnt/vol2/vol1/foo`. It is not a mount point, it resides on mount /mnt/vol2 with device 0:38 and TrimPrefix (=`rootPath`) is /vol1/foo. The loop will find a mount with this `rootPath` and device major/minor, although the directory is not mounted anywhere. I acknowledge that this example is very synthetic though. IMO, the rootPath of `/mnt/vol2` should be taken into account.

jsafrane replied:
It seems you fixed that in the meantime,sorry for the noise.

Mounting is hard and deciphering /proc/self/mountinfo even harder :-(

cofyc replied:
Yes, too hard...
Your example is added in test: https://github.com/kubernetes/kubernetes/pull/62102/commits/d6b4b12f004cb2844d774c60335bc2534f5d30cb.

cofyc commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1171,3 +1211,57 @@ func doSafeOpen(pathname string, base string) (int, error) {
+		} else if strings.HasPrefix(hostSource, mis[i].mountPoint) {
+			// If it's under a mount point.
+			mountID = mis[i].id
+			rootPath = filepath.Join("/", strings.TrimPrefix(hostSource, mis[i].mountPoint))
```
Updated in https://github.com/kubernetes/kubernetes/commit/ef50df7f933a9c349ce4eb4678be31e71b15c7eb.

cofyc commented:
/test pull-kubernetes-e2e-gce

cofyc commented:
/test pull-kubernetes-e2e-gce

cofyc commented:
/test pull-kubernetes-e2e-gce

cofyc commented:
/test pull-kubernetes-e2e-gce

cofyc commented:
hi, @msau42 @jsafrane 
Rebased and squashed, please take a look at again.

jsafrane commented:
/lgtm

Thanks for your patience.
